### ISSUE TITLE
Unexpected random crashes with java.lang.IllegalStateException: Unbalanced enter/exit

### ISSUE THREAD

**Issue by valdvit** [square/okhttp#8232]
2 weeks ago, we started experiencing spontaneous crashes after 5 to 10 minutes of app usage. No prior changes were implemented to the app before the crashes started. The logs indicate that the issue is related to the OkHttp/Okio library. This problem occurs on all devices. #3641 described a similar issue, it is suggested that it could be related to the D8 compiler, and a possible solution is to disable D8 in gradle.properties. However, in order to do this, I will have to downgrade Gradle to version 3.0.0, which was released in 2017. Are there any other possible solutions or workarounds for this?

Btw, we use OkHttp 4.10.0. 

```
FATAL EXCEPTION: OkHttp Dispatcher
java.lang.IllegalStateException: Unbalanced enter/exit
		at okio.AsyncTimeout$Companion.e(AsyncTimeout.kt:225)
		at okio.AsyncTimeout$Companion.b(AsyncTimeout.kt:201)
		at okio.AsyncTimeout.u(AsyncTimeout.kt:54)
		at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:144)
		at okio.RealBufferedSource.W(RealBufferedSource.kt:322)
		at okio.RealBufferedSource.a0(RealBufferedSource.kt:197)
		at okhttp3.internal.http1.HeadersReader.b(HeadersReader.kt:29)
		at okhttp3.internal.http1.Http1ExchangeCodec.d(Http1ExchangeCodec.kt:178)
		at okhttp3.internal.connection.Exchange.q(Exchange.kt:106)
		at okhttp3.internal.http.CallServerInterceptor.a(CallServerInterceptor.kt:79)
		at okhttp3.internal.http.RealInterceptorChain.a(RealInterceptorChain.kt:109)
		at okhttp3.internal.connection.ConnectInterceptor.a(ConnectInterceptor.kt:34)
		at okhttp3.internal.http.RealInterceptorChain.a(RealInterceptorChain.kt:109)
		at okhttp3.internal.cache.CacheInterceptor.a(CacheInterceptor.kt:95)
		at okhttp3.internal.http.RealInterceptorChain.a(RealInterceptorChain.kt:109)
		at okhttp3.internal.http.BridgeInterceptor.a(BridgeInterceptor.kt:83)
		at okhttp3.internal.http.RealInterceptorChain.a(RealInterceptorChain.kt:109)
		at okhttp3.internal.http.RetryAndFollowUpInterceptor.a(RetryAndFollowUpInterceptor.kt:76)
		at okhttp3.internal.http.RealInterceptorChain.a(RealInterceptorChain.kt:109)
		at okhttp3.internal.connection.RealCall.q(RealCall.kt:201)
		at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517)
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
		at java.lang.Thread.run(Thread.java:1012)
```

**Comment by swankjesse** [comment]
what changed 2 weeks ago?

Wanna upgrade to the latest OkHttp release? The 5.x code is stable; only the APIs are incomplete.

**Comment by valdvit** [comment]
@swankjesse Nothing changed at all. Neither on the Android side, nor on the server.

We would definitely upgrade to a newer release, but it pulls other dependencies like the Kotlin version and requires a huge amount of work to update the whole project. We're looking for a possible workaround just to stabilize our production which, has been working correctly for more than 4 months.

**Comment by yschimke** [comment]
There are bug fixes including for websocket race conditions in 4.12.0 that definitely could be relevant.  So it's probably worth checking if 4.12.0 fixes it.

**Comment by valdvit** [comment]
@yschimke Thank you, I’ll try to update to 5.x.x or 4.12.0 and tell the result.

**Comment by swankjesse** [comment]
Note that you can depend on OkHttp 5.x without changing your own project’s Kotlin compiler version. The newer Kotlin stdlib interops fine with code compiled with the older compiler.

**Comment by verticelo** [comment]
We have seen this issue as well since about the 19th of January for the React Native 0.72 version for an app with many users. Nothing was changed on the server side nor client side and all of a sudden we started to get a flood of crashes.

Yes yes, I know how it sounds that "nothing was changed". We didn't at least push any new application code but we have automatic OS updates on all the Ubuntu servers which might have caused something to change, alternatively AWS load balancers etc might have changed or OS updates on Android.

We will try to upgrade to a newer OkHttp to see if that resolves it but it would be nice to understand what actually caused it.

```
Fatal Exception: java.lang.IllegalStateException: Unbalanced enter/exit
       at okio.AsyncTimeout$Companion.scheduleTimeout(AsyncTimeout.kt:225)
       at okio.AsyncTimeout$Companion.access$scheduleTimeout(AsyncTimeout.kt:201)
       at okio.AsyncTimeout.enter(AsyncTimeout.kt:54)
       at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:331)
       at okio.RealBufferedSource.indexOf(RealBufferedSource.kt:427)
       at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.kt:320)
       at okhttp3.internal.http1.HeadersReader.readLine(HeadersReader.kt:29)
       at okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.kt:178)
       at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.kt:106)
       at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:79)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
       at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
       at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
       at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
       at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
       at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
       at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
       at java.lang.Thread.run(Thread.java:1012)
```

**Comment by yschimke** [comment]
There were some race condition bugs now fixed in 4.12.0. one is relevant if you use web sockets.

Otherwise it's likely to be a threading issue in your app, hopefully not since it was working before.

Server changes including any intermediate http proxies you have may change conditions so you now hit this. So client changes are not required to change the code path.

**Comment by valdvit** [comment]
@yschimke @swankjesse 
Updating from version 4.10.0 to 5.0.0-alpha.12 resolved the issue, no more crashes appear. Thank you for your help!
